docs/tla-plus/LeaderLeases/RaftLeader.tla

```diff
@@ -0,0 +1,661 @@
+  LeaderExclusivity == \A i, j \in Nodes :
+    raft[i].term = raft[j].term /\ raft[i].lead /= 0 /\ raft[j].lead /= 0 =>
+      raft[i].lead = raft[j].lead
+end define;
+
+\*************************************\
+\* Macros for Raft state transitions *\
+\*************************************\
+
+macro become_follower(term, lead)
+begin
+  if raft[self].term /= term then
+    raft[self].state := StateFollower ||
+    raft[self].term  := term          ||
+    raft[self].vote  := 0             ||
+    raft[self].lead  := lead;
+  else
+    raft[self].state := StateFollower ||
+    raft[self].lead  := lead;
+  end if;
+end macro;
+
+macro become_pre_candidate()
+begin
+  raft[self].state     := StatePreCandidate ||
+  raft[self].lead      := 0                 ||
+  raft[self].voteResps := {self};
+end macro;
+
+macro become_candidate()
+begin
+  raft[self].state     := StateCandidate      ||
+  raft[self].term      := raft[self].term + 1 ||
+  raft[self].vote      := self                ||
+  raft[self].lead      := 0                   ||
+  raft[self].voteResps := {self};
+end macro;
+
+macro become_leader()
+begin
+  raft[self].state    := StateLeader ||
+  raft[self].lead     := self        ||
+  raft[self].appResps := {self}      ||
+  raft[self].appended := FALSE;
+end macro;
+
+macro restart()
+begin
+  raft[self] := [
+    state     |-> StateFollower,
+    term      |-> raft[self].term, \* persistent
+    vote      |-> raft[self].vote, \* persistent
+    lead      |-> 0,
+    voteResps |-> {},
+    appResps  |-> {},
+    appended  |-> FALSE
+  ];
+end macro;
+
+\************************************\
+\* Macros for sending Raft messages *\
+\************************************\
+
+macro send_msg(to, msg)
+begin
+  network[to] := Append(network[to], msg);
+end macro
+
+macro send_msg_pre_vote(to)
+begin
+  send_msg(to, [
+    type |-> MsgPreVote,
+    from |-> self,
+    term |-> raft[self].term + 1
+  ]);
+end macro
+
+macro send_msg_pre_vote_resp(to, reject)
+begin
+  send_msg(to, [
+    type   |-> MsgPreVoteResp,
+    from   |-> self,
+    term   |-> raft[self].term,
+    reject |-> reject
+  ]);
+end macro
+
+macro send_msg_vote(to)
+begin
+  send_msg(to, [
+    type   |-> MsgVote,
+    from   |-> self,
+    term   |-> raft[self].term
+  ]);
+end macro
+
+macro send_msg_vote_resp(to, reject)
+begin
+  send_msg(to, [
+    type   |-> MsgVoteResp,
+    from   |-> self,
+    term   |-> raft[self].term,
+    reject |-> reject
+  ]);
+end macro
+
+macro send_msg_app(to)
+begin
+  send_msg(to, [
+    type   |-> MsgApp,
+    from   |-> self,
+    term   |-> raft[self].term
+  ]);
+end macro
+
+macro send_msg_app_resp(to, reject)
+begin
+  send_msg(to, [
+    type   |-> MsgAppResp,
+    from   |-> self,
+    term   |-> raft[self].term,
+    reject |-> reject
+  ]);
+end macro
+
+macro campaign()
+begin
+  if UsePreVote then
+    become_pre_candidate();
+    with i \in Nodes \ {self} do
+      send_msg_pre_vote(i);
```

Comment:
Ah, this isn't a broadcast. `with` picks a single element rather than all. Are we relying here on the fact that there are 3 voters, so sending a message to 1 is enough?

Reply:
You're right, this is only picking a single element, which was not the intention. It should be a true broadcast, and then individual messages should be able to be delivered out-of-order.

Reply:
Fixed.

Reply:
FWIW, I confirmed from the TLA+ coverage trace that the 5x replication spec was never electing a leader before, but now it is.